File: Striver/Recursion/q20.java
package Striver.Recursion;

class q20 {
    // n-queens problem - backtracking

    public static void main(String[] args) {
        int n = 4;
        int[][] board = new int[n][n];
        solveNQueens(board, 0);
    }

    public static void solveNQueens(int[][] board, int col) {
        if (col == board.length) {
            printBoard(board);
            return;
        }

        for (int i = 0; i < board.length; i++) {
            if (isSafe(board, i, col)) {
                board[i][col] = 1;
                solveNQueens(board, col + 1);
                board[i][col] = 0;
            }
        }
    } 

    public static boolean isSafe(int[][] board, int row, int col) {

        // checking for left side same row
        int i = row, j = col;
        while (j >= 0) {
            if (board[i][j] == 1)
                return false;
            j--;
        }

        // checking for upper left diagonal
        i = row;
        j = col;
        while (i >= 0 && j >= 0) {
            if (board[i][j] == 1)
                return false;
            i--;
            j--;
        }

        //checking for lower left diagonal
        i = row;
        j = col;
        while (i < board.length && j >= 0) {
            if (board[i][j] == 1)
                return false;
            i++;
            j--;
        }

        return true;
    }

    public static void printBoard(int[][] board) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board.length; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
        System.out.println();
    }
}

File: Striver/Recursion/q20.py
# n-queen - backtracking

def nQueen(arr,col):
    if col >= len(arr):
        print(arr)
        return
    for i in range(len(arr)):
        if isSafe(arr,i,col):
            arr[i][col] = 1
            nQueen(arr,col+1)
            arr[i][col] = 0

def isSafe(arr,row,col):
    for i in range(col):
        if arr[row][i] == 1:
            return False
    for i,j in zip(range(row,-1,-1),range(col,-1,-1)):
        if arr[i][j] == 1:
            return False
    for i,j in zip(range(row,len(arr),1),range(col,-1,-1)):
        if arr[i][j] == 1:
            return False
    return True


arr = [[0 for i in range(4)] for j in range(4)]
nQueen(arr,0)
